client/src/store/historyStore/model/watchHistory.js

```diff
@@ -19,46 +20,134 @@ let lastUpdateTime = null;
 
+/**
+ * Periodically watches for changes in the current History
+ */
 export async function watchHistory() {
-    // get current history
     const history = await getCurrentHistoryFromServer();
-    const historyId = history.id;
 
-    // continue if the history update time has changed
-    if (!lastUpdateTime || lastUpdateTime < history.update_time) {
+    if (historyHasChanges(history)) {
+        console.debug("History changes detected");
         lastUpdateTime = history.update_time;
-        // execute request to obtain recently changed items
-        const params = {
-            limit: limit,
-            q: "update_time-ge",
-            qv: lastRequestDate.toISOString(),
-            v: "dev",
-            view: "detailed",
-        };
-        const paramsString = Object.entries(params)
-            .map(([k, v]) => `${k}=${encodeURIComponent(v)}`)
-            .join("&");
-        const url = `api/histories/${historyId}/contents?${paramsString}`;
-        lastRequestDate = new Date();
-        const payload = await urlData({ url });
-        // show warning that not all changes have been obtained
-        if (payload && payload.length == limit) {
-            console.debug(`Reached limit of monitored changes (limit=${limit}).`);
-        }
-        // pass changed items to attached stores
-        store.commit("history/setHistory", history);
-        store.commit("saveDatasets", { payload });
-        store.commit("saveHistoryItems", { historyId, payload });
-        store.commit("saveCollectionObjects", { payload });
-        // trigger changes in legacy handler
-        const Galaxy = getGalaxyInstance();
-        if (Galaxy) {
-            Galaxy.user.fetch({
-                url: `${Galaxy.user.urlRoot()}/${Galaxy.user.id || "current"}`,
-            });
-        }
+        await requestHistoryChanges(history);
     }
+
     setTimeout(() => {
         watchHistory();
     }, throttlePeriod);
 }
+
+function historyHasChanges(history) {
+    return !lastUpdateTime || lastUpdateTime < history.update_time;
+}
+
+/**
+ * Requests history changes since the last request in two rounds.
+ * - The first round only retrieves changes affecting the currently filtered items and updates
+ * the query counts appropriately.
+ * - The second round fetches all the changes in background.
+ * - Both queries will update the appropriate stores if there are changes.
+ * @param {*} history The current history
+ */
+async function requestHistoryChanges(history) {
+    const historyId = history.id;
+    const since = lastRequestDate.toISOString();
+    lastRequestDate = new Date();
+    // First round: request changes affecting currently filtered items and update query stats
+    const data = await requestHistoryChangesWithStatsSince(historyId, since);
+    handleHistoryChanges(history, data);
+    // Second round: request possible additional changes in background
+    requestAllHistoryChangesSince(historyId, since).then((changes) => {
```

Comment:
The only alternative I can think of right now is to merge both queries in the same request.

Maybe move this kind of response to a new API route like:

```
GET api/histories/{history_id}/contents/changes?since={date}&{filtering_params}
```
This new route can just call the same service method as the index with the right parameters and possibly make the `since` parameter required, probably making more clear the intent.

And modifying the response to something like this:

```jsonc
{
  "since": {date}, // The same date used in the request
  "stats": {
    total_matches: X, // The number of items matching the {filtering_params}
    total_changes: Y, // The number of items updated since {date} ignoring the {filtering_params}
  },
  "changes": [
    // contains all changed items since {date}, ignoring the {filtering_params}
  ]
}
```

We can then paginate the changes if necessary using the `since` and the `total_changes` as of right now the number of changes after a bulk operation can be huge and we don't paginate them.

Reply:
These changes are no longer part of this PR, we can follow up with the watch enhancements in a different PR.